### ISSUE TITLE
dhall incorrectly(?) reports a record with a wrong field name as an internal type error

### ISSUE THREAD

**Issue by googleson78** [dhall-lang/dhall-haskell#975]
^title

Repro files:

`A.dhall`
```haskell
{ field : Text }
```
`B.dhall`
```haskell
let f = \(x : ./A.dhall) -> x.field
in f { not-the-same-field = "" }
```

And result:
```
tauren :: /tmp/dhall » dhall --annotate --explain <<< './B.dhall'
Error: Compiler bug

An ill-typed expression was encountered during normalization.
Explanation: This error message means that there is a bug in the Dhall compiler.
You didn't do anything wrong, but if you would like to see this problem fixed
then you should report the bug at:

https://github.com/dhall-lang/dhall-haskell/issues

CallStack (from HasCallStack):
  error, called at src/Dhall/Eval.hs:533:44 in dhall-1.23.0-ea4449817018388a6113eb08067bde5d09d62c716293d97ade5ba0d0c6dc3a24:Dhall.Eval
```

**Comment by Gabriel439** [comment]
This is the same issue as https://github.com/dhall-lang/dhall-haskell/issues/959, which is fixed in `master`.  The issue here is that the normalization code throws an `error` if it is asked to normalize an ill-typed expression and due to that bug some ill-typed expressions were not properly rejected by the type-checker.  Fixing the type-checker catches the error with a proper type error instead of a normalization-time internal error.

I verified that your example indeed fails using the latest `dhall` version on Hackage, but is fixed by `master`:

```
$ dhall <<< './example.dhall'

Use "dhall --explain" for detailed errors

↳ ./example.dhall

f : ∀(x : { field : Text }) → Text

Error: Wrong type of function argument

{ - field : …
, + not-the-same-field : …
}

2:     f { not-the-same-field = "" }

/home/gabriel/proj/dhall-haskell/dhall/example.dhall:2:5
```
I will cut a release in the next few days to get the fix out and leave this issue open until then.

**Comment by Gabriel439** [comment]
The fix is released now as `dhall-1.24.0`
